File: Final_Project/Code/Project_4/Bai_toan_5/Exercises/2/CSS_bt5_ex2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>
using namespace std;

class SGBGraph {
public:
    int n, m;                       // so dinh, so canh
    vector<vector<int>> adj;        // danh sach ke
    vector<string> labels;          // nhan cua cac dinh

    SGBGraph() : n(0), m(0) {}

    void read_sgb(const string& filename) {
        // Doc file SGB de xay dung do thi
        ifstream file(filename);
        string line;
        getline(file, line);
        stringstream ss(line);
        string temp;
        ss >> temp >> temp >> temp >> n >> m;  // Parse header
        adj.assign(n, {});
        labels.assign(n, "");

        bool vertex_section = false, arc_section = false;
        while (getline(file, line)) {
            if (line.find("* Vertices") != string::npos) {
                vertex_section = true;
                continue;
            }
            if (line.find("* Arcs") != string::npos) {
                vertex_section = false;
                arc_section = true;
                continue;
            }
            if (vertex_section) {
                stringstream ss2(line);
                int v;
                string label;
                ss2 >> v >> label;
                labels[v] = label;
            }
            if (arc_section) {
                stringstream ss2(line);
                int v, u_idx;
                ss2 >> v >> u_idx;
                for (int i = 0; i < n; ++i) {
                    if (labels[i] == to_string(u_idx)) {
                        adj[i].push_back(v);
                        break;
                    }
                }
            }
        }
        file.close();
    }

    void write_sgb(const string& filename) {
        // Ghi do thi ra file SGB
        ofstream file(filename);
        file << "* GraphBase graph(utiltypes..., " << n << ", " << m << ")\n";
        file << "Graph description\n";
        file << "* Vertices\n";
        for (int i = 0; i < n; ++i) {
            file << i << ", v" << i << ", 0, 0\n";
        }
        file << "* Arcs\n";
        int edge_id = 0;
        for (int u = 0; u < n; ++u) {
            for (int v : adj[u]) {
                file << v << ", " << edge_id << ", 0, 0\n";
                edge_id++;
            }
        }
        file << "* Checksum\n";
        file.close();
    }
};

int main() {
    SGBGraph g;
    // g.read_sgb("input.txt");
    // g.write_sgb("output.txt");
    return 0;
}

File: Final_Project/Code/Project_4/Bai_toan_5/Exercises/2/CSS_bt5_ex2.py
class SGBGraph:
    def __init__(self):
        # n: so dinh
        # m: so canh
        # adj: danh sach ke
        # labels: nhan cua cac dinh
        self.n = 0
        self.m = 0
        self.adj = []
        self.labels = []

    def read_sgb(self, filename):
        # Doc file SGB de xay dung do thi
        with open(filename, 'r') as f:
            lines = f.readlines()
        header = lines[0].strip().split()
        self.n = int(header[3][:-1])
        self.m = int(header[4][:-1])
        self.adj = [[] for _ in range(self.n)]
        self.labels = [''] * self.n

        vertex_section = False
        arc_section = False
        for line in lines[1:]:
            line = line.strip()
            if line.startswith('* Vertices'):
                vertex_section = True
                continue
            if line.startswith('* Arcs'):
                vertex_section = False
                arc_section = True
                continue
            if vertex_section:
                parts = line.split(',')
                v = int(parts[0])
                self.labels[v] = parts[1].strip()
            if arc_section:
                parts = line.split(',')
                v = int(parts[0])
                u = int(self.labels.index(parts[1].strip()))
                self.adj[u].append(v)

    def write_sgb(self, filename):
        # Ghi do thi ra file SGB
        with open(filename, 'w') as f:
            f.write(f"* GraphBase graph(utiltypes..., {self.n}, {self.m})\n")
            f.write("Graph description\n")
            f.write("* Vertices\n")
            for i in range(self.n):
                f.write(f"{i}, v{i}, 0, 0\n")
            f.write("* Arcs\n")
            edge_id = 0
            for u in range(self.n):
                for v in self.adj[u]:
                    f.write(f"{v}, {edge_id}, 0, 0\n")
                    edge_id += 1
            f.write("* Checksum\n")

# Vi du su dung
if __name__ == "__main__":
    g = SGBGraph()
    # g.read_sgb("input.txt")
    # g.write_sgb("output.txt")